src/typeInference/annotateBinaryOperator.ts

```diff
@@ -0,0 +1,47 @@
+import { SequenceType, sequenceTypeToString, ValueType } from '../expressions/dataTypes/Value';
+import {
+	generateBinaryOperatorType,
+	getBinaryPrefabOperator,
+} from '../expressions/operators/arithmetic/BinaryOperator';
+import astHelper, { IAST } from '../parsing/astHelper';
+
+/**
+ * Annotate the binary operators on the numeric and date types
+ * by inserting a new attribute `type` to the AST.
+ *
+ * @param ast the AST to be annotated.
+ * @param left the left hand side operand.
+ * @param right the right hand side operand.
+ * @param operator addOp, subtractOp, divOp, idivOp, modOp, multiplyOp.
+ * @throws Error when multiplicities of the operands don't match.
+ * @throws Error XPTY0004 when operators cannot be applied on operands.
+ * @returns the inferred type or `undefined` when missing operand(s).
+ */
+export function annotateBinOp(
+	ast: IAST,
+	left: SequenceType | undefined,
+	right: SequenceType | undefined,
+	operator: string
+): SequenceType | undefined {
+	// If we don't have the left and right type, we cannot infer the current type
+	if (!left || !right) {
+		return undefined;
+	}
+
+	// TODO: Fix this hack (pathExpr returns a node in 1 case, which cannot be added to an integer)
+	if (left.type === ValueType.NODE || right.type === ValueType.NODE) return undefined;
```

Comment:
Friendly reminder. return type is still a XSNUMERIC, right? The Node cannot be added, but it can go trough a process of `atomization`, which renders it a `xs:anyAtomicType` iirc, which can be added to a number if it matches the xs:integer / xs:float / xs:decimal productions

Reply:
Thanks for the feedback and it has been resolved by 2c530a834da44cc7beb9450195a6700d79a19843